server/app/repository.py

```diff
@@ -75,6 +77,19 @@ class StationPhoto(TypedDict):
     filename: str
     descr: str
 
+class UserRole(Enum):
```

Comment:
Check for possibility use string or Literal (better) for it. Maybe you avoid conversion from int to string and from string to int when fetch data from DB. In dynamic typed languages as Python Enums have less advantages then static typed languages.

Reply:
For some reason there are two aspects here in this thread.

The first one is about read_user. I've updated the code to use single union parameter.

Reply:
Regarding, the second aspect (replace Enum with Literals), here's my answer:

There is no conversion from integers (at least not in our code. Perhaps somewhere deep in psycopg2, but I don't care). The underlying enum type in postgres is returned as a string and this is what cursor.fetchone() returns. I can get rid of the Enum and replace it with a class that would have 4 literals defined, but I don't understand why it would be any better. I can iterate over Enum to get all values. This may be useful in the future to display some help information, sanity check admin panel data etc.

Can you point me to an example class somewhere that uses Literals as enum values? I don't understand what exaclty you propose here. I want something that would accept only explicit 4 values and would throw for everything else. I find it useful if it's possible to programmatically get a list of all possible values. Having explicit expression in the code everywhere values are used (e.g. you grep for UserRole.ADMIN and you find ALL references) is another big bonus.

Reply:
I've tried to experiment with Literal, but after a while I strongly reject the idea as strongly inferior to Enum. Here are my reasons:

1. You can't iterate over Literal. There is a good reason for that. Soon we'll be implementing an admin panel and we'll need a list of all possible roles.
2. I still haven't heard any convincing argument why to use Literal at all.
3. I did an experiment and wanted to try Literal in practice. I've modified the UserRole(Enum) with UserRole = Literal[...] and got this exception. I must admit it's confusing as hell for a simple enum.

```
app/repository.py:397: in user_role_to_enum
    u = first(lambda u: u.name == role.upper(), UserRole)
app/utils.py:28: in first
    for item in items:
/usr/lib/python3.7/typing.py:254: in inner
    return func(*args, **kwds)
/usr/lib/python3.7/typing.py:629: in _getitem_
    params = tuple(_type_check(p, msg) for p in params)
/usr/lib/python3.7/typing.py:629: in <genexpr>
    params = tuple(_type_check(p, msg) for p in params)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

arg = 0, msg = 'Parameters to generic types must be types.', is_argument = True

    def _type_check(arg, msg, is_argument=True):
        """Check that the argument is a type, and return it (internal helper).
    
        As a special case, accept None and return type(None) instead. Also wrap strings
        into ForwardRef instances. Consider several corner cases, for example plain
        special forms like Union are not valid, while Union[int, str] is OK, etc.
        The msg argument is a human-readable error message, e.g::
    
            "Union[arg, ...]: arg should be a type."
    
        We append the repr() of the actual value (truncated to 100 chars).
        """
        invalid_generic_forms = (Generic, _Protocol)
        if is_argument:
            invalid_generic_forms = invalid_generic_forms + (ClassVar, )
    
        if arg is None:
            return type(None)
        if isinstance(arg, str):
            return ForwardRef(arg)
        if (isinstance(arg, _GenericAlias) and
                arg.__origin__ in invalid_generic_forms):
            raise TypeError(f"{arg} is not valid as type argument")
        if (isinstance(arg, _SpecialForm) and arg not in (Any, NoReturn) or
                arg in (Generic, _Protocol)):
            raise TypeError(f"Plain {arg} is not valid as type argument")
        if isinstance(arg, (type, TypeVar, ForwardRef)):
            return arg
        if not callable(arg):

              raise TypeError(f"{msg} Got {arg!r:.100}.")

E           TypeError: Parameters to generic types must be types. Got 0.

/usr/lib/python3.7/typing.py:142: TypeError
```

4. Continuing with my experiment, I've removed user_role_to_enum and all references to it, and replaced UserRole.ADMIN with 'admin' literal. Now the tests started to fail:
```
E       AssertionError: 'admin' != 'ADMIN'
E       - admin
E       + ADMIN
``` 
Which is a proof that in the process we also lost value checking in the process. Yes, this could be solved by using capital 'ADMIN' everywhere, but this puts the burden of remembering if it's capital or lowercase on user. I was using Visual Studio Code. The env did not stop me and didn't give any hints.

5. Literal is a recent addition. It's in typing in python 3.8, but in typing_extensions in 3.7. Don't know if it's available in 3.6.

As such, I discarded my experimental changes and strongly feel that going forward with Enum is the way to go.